src/backend/joanie/payment/models.py

```diff
@@ -1 +1,465 @@
+
+from joanie.core.models import Order
+from joanie.core.utils import merge_dict
+
+from . import enums as payment_enums
+from . import get_payment_backend
+
+User = get_user_model()
+
+
+class Invoice(models.Model):
+    """
+    Invoice model is an accounting element related to an order
+    """
+
+    parent = models.ForeignKey(
+        "self",
+        on_delete=models.RESTRICT,
+        related_name="children",
+        verbose_name="parent",
+        blank=True,
+        null=True,
+    )
+    reference = models.CharField(
+        db_index=True,
+        editable=False,
+        max_length=40,
+        unique=True,
+    )
+    created_on = models.DateTimeField(
+        _("created on"),
+        auto_now_add=True,
+        editable=False,
+    )
+    updated_on = models.DateTimeField(
+        _("updated on"),
+        auto_now=True,
+        editable=False,
+    )
+    order = models.OneToOneField(
+        Order,
+        verbose_name=_("order"),
+        on_delete=models.PROTECT,
+        blank=True,
+        null=True,
+    )
+    total = MoneyField(
+        _("total"),
+        max_digits=9,
+        decimal_places=2,
+        default_currency=settings.DEFAULT_CURRENCY,
+        null=False,
+        blank=False,
+    )
+    recipient_name = models.CharField(
+        _("invoice recipient"), max_length=40, null=False, blank=False
+    )
+    recipient_address = models.TextField(
+        _("invoice address"), max_length=255, null=False, blank=False
+    )
+    localized_context = models.JSONField(
+        _("context"),
+        help_text=_("Localized data that needs to be frozen on invoice creation"),
+        editable=False,
+    )
+
+    class Meta:
+        db_table = "joanie_invoice"
+        verbose_name = _("Invoice")
+        verbose_name_plural = _("Invoices")
+        constraints = [
+            models.CheckConstraint(
+                check=(
+                    models.Q(order__isnull=False)
+                    & models.Q(parent__isnull=True)
+                    & models.Q(total__gte=0)
+                )
+                | (models.Q(order__isnull=True) & models.Q(parent__isnull=False)),
+                name="invoice_relies_on_order_or_invoice",
+            ),
+        ]
+
+    def __str__(self):
+        types = dict(payment_enums.INVOICE_TYPES)
+        return f"{types[self.type]} {self.reference}"
+
+    @property
+    def state(self):
+        """
+        Process the state of the invoice
+        """
+        if self.balance.amount >= 0:
+            if (
+                self.invoiced_balance.amount == 0
+                and self.transactions_balance.amount == 0
+            ):
+                return payment_enums.INVOICE_STATE_REFUNDED
+
+            return payment_enums.INVOICE_STATE_PAID
+
+        return payment_enums.INVOICE_STATE_UNPAID
+
+    @property
+    def related_order(self):
+        """
+        Return the related order.
+        """
+        return self.order or self.parent.order
```

Comment:
another way would be to always fill the order... it would allow to make the order field required and avoid bugs + avoid a join to get the order for a credit note?

Reply:
I liked the idea to know which is the "root" invoice and keep a OneToOne relation.

Reply:
Not sure what you meant is good: make the order required? We can always know which one is the root because invoice.parent is None?

Reply:
Right, I updated the code. For now, order property is a foreign key and each invoice is linked to an order.